web-api/src/storage/postgres/client.rs

```diff
@@ -0,0 +1,329 @@
+pub(crate) struct Database {
+    pool: Pool<Postgres>,
+    tenant_db_name: QuotedIdentifier,
+}
+
+impl Database {
+    #[instrument]
+    pub(crate) async fn build(
+        config: &Config,
+        enable_legacy_tenant: bool,
+    ) -> Result<DatabaseBuilder, SetupError> {
+        let options = Self::build_connection_options(config)?;
+        info!("starting postgres setup");
+        let pool = PoolOptions::new()
+            .after_release(|conn, _metadata| {
+                async {
+                    sqlx::query("RESET ROLE;").execute(conn).await?;
+                    Ok(true)
+                }
+                .boxed()
+            })
+            .connect_with(options)
+            .await?;
+
+        if !config.skip_migrations {
+            sqlx::migrate!().run(&pool).await?;
+
+            //FIXME: This will be handled nicer once schema migration has it's own
+            //       API the follow up PR
+            if enable_legacy_tenant {
+                Self::legacy_tenant_setup(&pool).await?;
+            }
+        }
+
+        Ok(DatabaseBuilder { pool })
+    }
+
+    #[instrument(err)]
+    async fn legacy_tenant_setup(pool: &Pool<Postgres>) -> Result<(), SetupError> {
+        //FIXME: THIS CODE WILL BE COMPLETELY REPLACED IN FOLLOW UP PR
+
+        let db_name = db_name_for_tenant_id(TenantId::missing())?;
+
+        let mut tx = pool.acquire().await?;
+        // this is a "global" db migration lock it will be held until
+        // - we call unlock or
+        // - the session ends, e.g. we return an error here and drop the connection pool
+        tx.lock().await?;
+
+        let role_exists =
+            sqlx::query::<Postgres>("SELECT FROM pg_catalog.pg_roles WHERE rolname = $1;")
+                .bind(db_name.unquoted_str())
+                .execute(&mut tx)
+                .await?
+                .rows_affected()
+                > 0;
+
+        //WARNING: This has issues with tests as tests have one db per test
+        //          but users are per db service not db instance.
+        //          This can make the test fail in two ways:
+        //          - role exist from test run on other branch but with different
+        //            permissions, if so clear test db
+        //          - test are run using multiple processes (not threads) and try
+        //            to create the user at the same time, in this case
+        //
+        //          There is a fix in the follow up PR but it can't be back ported.
+        if !role_exists {
+            // hack to work around cross db concurrency problems we only have
+            // There is probably a better way to do it, but it get's removed anyway.
+            static ROLE_CREATION_SYNC: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
+            let mut initialized = ROLE_CREATION_SYNC.lock().await;
+            if !*initialized {
+                let (current_user,) = sqlx::query_as::<_, (String,)>("SELECT CURRENT_USER;")
```

Comment:
many places in the postgres documentation use capital letters for `CURRENT_USER` and similar (e.g. `SESSION_USER`).  Maybe because while it is a function it can't be called with `()`.

Reply:
code is gone now :tada: